primitives/src/gateway_inbound_protocol.rs

```diff
@@ -0,0 +1,175 @@
+        return_value: Option<Vec<u8>>,
+    ) -> Result<CircuitOutboundMessage, &'static str>;
+
+    /// Call custom dispatchable call behind that gateway in a reversible (escrowed) way - enforcing read-pnly operations. Returns (gets it delivered by relayers):
+    /// stamp - Event generated on that chain
+    /// stamp_proof - Merkle Path in storage trie (that's where the stamp lands) and block reference
+    /// finality_proof - Proof that the block is finalized
+    fn custom_call_escrow(
+        &self,
+        module_name: &str,
+        fn_name: &str,
+        data: Vec<u8>,
+        to: Vec<u8>,
+        value: Vec<u8>,
+        gas: Vec<u8>,
+        gateway_type: GatewayType,
+        return_value: Option<Vec<u8>>,
+    ) -> Result<CircuitOutboundMessage, &'static str>;
+
+    /// Transfer balance on a chain behind that gateway in a reversible (escrowed) way. Returns (gets it delivered by relayers):
+    /// stamp - Event generated on that chain
+    /// stamp_proof - Merkle Path in storage trie (that's where the stamp lands) and block reference
+    /// finality_proof - Proof that the block is finalized
+    fn transfer(
+        &self,
+        to: GenericAddress,
+        value: Compact<u128>,
+        gateway_type: GatewayType,
+    ) -> Result<CircuitOutboundMessage, &'static str>;
+
+    /// Transfer balance on a chain behind that gateway in a dirty way - without possibility to revert that action. Returns (gets it delivered by relayers):
+    /// stamp - Event generated on that chain
+    /// stamp_proof - Merkle Path in storage trie (that's where the stamp lands) and block reference
+    /// finality_proof - Proof that the block is finalized
+    fn transfer_escrow(
+        &self,
+        escrow_account: Vec<u8>,
+        requester: Vec<u8>,
+        to: Vec<u8>,
+        value: Vec<u8>,
+        transfers: &mut Vec<TransferEntry>,
+        gateway_type: GatewayType,
+    ) -> Result<CircuitOutboundMessage, &'static str>;
+
+    /// Swap X tokens to Y different tokens on a chain behind that gateway in a dirty way - without possibility to revert that action. Returns (gets it delivered by relayers):
+    /// stamp - Event generated on that chain
+    /// stamp_proof - Merkle Path in storage trie (that's where the stamp lands) and block reference
+    /// finality_proof - Proof that the block is finalized
+    fn swap_dirty(
+        &self,
+        to: Vec<u8>,
+        value: Vec<u8>,
+        gas: Vec<u8>,
+        gateway_type: GatewayType,
+    ) -> Result<CircuitOutboundMessage, &'static str>;
+
+    /// Swap X tokens to Y different tokens on a chain behind that gateway in a reversible (escrowed) way. Returns (gets it delivered by relayers):
+    /// stamp - Event generated on that chain
+    /// stamp_proof - Merkle Path in storage trie (that's where the stamp lands) and block reference
+    /// finality_proof - Proof that the block is finalized
+    fn swap_escrow(
+        &self,
+        from: Vec<u8>,
```

Comment:
We should make args associate type so that we dont have to worry about decoding and encoding back.